### ISSUE TITLE
TileLayer's ZoomOffset option don't work in 1.3.0

### ISSUE THREAD

**Issue by lpikora** [Leaflet/Leaflet#6004]
## How to reproduce

- Leaflet version I'm using: 1.3.0
- Browser (with version) I'm using: Chrome
- OS/Platform (with version) I'm using: MacOs
- Add any TileLayer to map.
- Set zoomOffset option to some value

## What behaviour I'm expecting and which behaviour I'm seeing

Tile request urls have zoom + offset.

## Minimal example reproducing the issue

- [x] this example is as simple as possible
- [x] this example does not rely on any third party code

```
Mapbox: L.tileLayer('//api.mapbox.com/styles/v1/mapbox/streets-v9/tiles/{z}/{x}/{y}?access_token=' + token, {
                  attribution: '&copy; <a href="https://www.mapbox.com/about/maps/">Mapbox</a> &copy; <a href="http://www.openstreetmap.org/copyright">OpenStreetMap</a>',
                  tileSize: 512,
                  zoomOffset: -10,
                  detectRetina: true,
                  maxZoom: 21
              }),
```

i think that problem is in `z: coords.z ? coords.z : this._getZoomForUrl()` in `getTileUrl` method.

`this._getZoomForUrl()` is never reached...

```
	// @section Extension methods
	// @uninheritable
	// Layers extending `TileLayer` might reimplement the following method.
	// @method getTileUrl(coords: Object): String
	// Called only internally, returns the URL for a tile given its coordinates.
	// Classes extending `TileLayer` can override this function to provide custom tile URL naming schemes.
	getTileUrl: function (coords) {
		var data = {
			r: retina ? '@2x' : '',
			s: this._getSubdomain(coords),
			x: coords.x,
			y: coords.y,
			z: coords.z ? coords.z : this._getZoomForUrl()
		};
		if (this._map && !this._map.options.crs.infinite) {
			var invertedY = this._globalTileRange.max.y - coords.y;
			if (this.options.tms) {
				data['y'] = invertedY;
			}
			data['-y'] = invertedY;
		}

		return template(this._url, extend(data, this.options));
	},

```

**Comment by cherniavskii** [comment]
From a fast look, it may be fixed like that:

```diff
- z: coords.z ? coords.z : this._getZoomForUrl()
+ z: coords.z ? coords.z + this.options.zoomOffset : this._getZoomForUrl()
```

I'm wondering if `zoomReverse` option should be handled to work with zoom parameter as well? :thinking:

**Comment by ghybs** [comment]
Hi,

Hum it seems to me that this bug would need first to give a closer look at PR #5822, as spotted by @lpikora, before trying to patch it over…

That PR is not linked to a proper issue.
It has little explanation on why the modification is fixing a bug, or how it would improve the code for the general situation.
Rather, it says that it is a port of an override for a particular project.
And obviously it does not have any associated test.

It only modified the line:
https://github.com/Leaflet/Leaflet/blob/0f3aed37b90cce8f0207136ddb4d660fe58a01c0/src/layer/tile/TileLayer.js#L167

It is true that this line might look confusing:
`coords.z` is previously assigned, and here (i.e. before the PR) it is purely replaced by a new value (calling `_getZoomForUrl()`) that does not even use `coords.z` as an input.
Hence a quick review might conclude that there is something missing, and that the PR might be relevant.

However, the replacement of `coords.z` seems to be on purpose: it is initially used as internal value for the Grid Layer, then it is replaced for external URL usage.
It is indeed within the call to [`_getZoomForUrl`](https://github.com/Leaflet/Leaflet/blob/1a38558f7b22f5ffacda5f6a1b8e60d54c361873/src/layer/tile/TileLayer.js#L201-L212) that options (like `zoomOffset`) are applied.
Now we could argue about why modifying an input parameter, instead of cloning. But that is beyond the current bug and very probably beyond the initial intention of that PR.

Therefore to me this bug is rather challenging PR #5822:
is the particular project, mentioned in that PR, representative of the general usage?
Enough that the proposed modification also benefits others?

From a quick look to that project (or plugin?), since it is the main clue we have to justify the PR, it dates from 4 years ago with an update 3 months ago (which triggered the mentioned PR).
The Leaflet version originally used was `0.7.2`, where Tile Layer used `_loadTile` method.

I do not know whether it has been tested on Leaflet 1.0.0+ or not, and we can also challenge whether the mentioned method override is appropriate or not.

Therefore IMO resolving this bug should be done by reverting that PR first.

**Comment by petrot** [comment]
In my app, the detectRetina = true causes the display problem.

**Comment by cherniavskii** [comment]
@ghybs Thanks for investigating this! I agree with you - that PR should be reverted

**Comment by cherniavskii** [comment]
@petrot if you experience issues with Leaflet - please submit an issue with detailed description

**Comment by cherniavskii** [comment]
Closed by #6006

**Comment by cherniavskii** [comment]
Fix available in Leaflet 1.3.1 http://leafletjs.com/2018/01/18/leaflet-1.3.1.html
